### PR TITLE
DT-235 Implement the write to Bed Assignment Histories

### PR DESCRIPTION
**Pull request by mikehalmamoj**
This is still WIP as there is more to do.  This PR can be merged into master as the endpoint still does nothing (yet).

### CODE REVIEW

File: src/main/java/net/syscon/elite/repository/jpa/model/BedAssignmentHistory.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        @Column(name = "BED_ASSIGN_SEQ", nullable = false)
+        private Integer sequence;
+    }
+
+    @EmbeddedId
+    private BedAssignmentHistoryPK bedAssignmentHistoryPK;
+
+    @ManyToOne
+    @JoinColumn(name = "OFFENDER_BOOK_ID", insertable = false, updatable = false)
+    private OffenderBooking offenderBooking;
```
Comment (mikehalmamoj) [inline src/main/java/net/syscon/elite/repository/jpa/model/BedAssignmentHistory.java]: This was required to handle the foreign key on offender booking id, which is part of the composite key and so cannot handle the join.

File: src/main/java/net/syscon/elite/repository/jpa/repository/BedAssignmentHistoriesRepository.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package net.syscon.elite.repository.jpa.repository;
+
+import net.syscon.elite.repository.jpa.model.BedAssignmentHistory;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.CrudRepository;
+import org.springframework.data.repository.query.Param;
+
+public interface BedAssignmentHistoriesRepository extends CrudRepository<BedAssignmentHistory, BedAssignmentHistory.BedAssignmentHistoryPK> {
+
+    @Query("select coalesce(max(b.bedAssignmentHistoryPK.sequence), 0) from BedAssignmentHistory b where b.bedAssignmentHistoryPK.offenderBookingId = :bookingId")
```
Comment (mikehalmamoj) [inline src/main/java/net/syscon/elite/repository/jpa/repository/BedAssignmentHistoriesRepository.java]: Considered a findFirstBy....Desc, but this way I can default to 0 if there are no records.

File: src/main/java/net/syscon/elite/service/LocationService.java [hunk 3]
```diff
@@ -94,11 +94,6 @@ public Location getLocation(final long locationId) {
         return locationRepository.findLocation(locationId).orElseThrow(EntityNotFoundException.withId(locationId));
     }
 
-    @Transactional
-    public void addBedAssignmentHistory(final Long bookingId, final Long livingUnitId) {
```
Comment (mikehalmamoj) [inline src/main/java/net/syscon/elite/service/LocationService.java]: This ended up in the BedAssignmentHistoryService

File: src/test/java/net/syscon/elite/service/BedAssignmentHistoryServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        final var now = LocalDateTime.now();
+        service.add(1L, 2L, "RSN", now);
+
+        verify(repository).save(argThat(bedAssignment ->
+                bedAssignment.getBedAssignmentHistoryPK().getOffenderBookingId() == 1L
+                        && bedAssignment.getLivingUnitId() == 2L
+                        && bedAssignment.getAssignmentReason().equals("RSN")
+                        && bedAssignment.getAssignmentDate().isEqual(now.toLocalDate())
+                        && bedAssignment.getAssignmentDateTime().isEqual(now)));
+    }
```
Comment (mikehalmamoj) [inline src/test/java/net/syscon/elite/service/BedAssignmentHistoryServiceTest.java]: I'm pining for mockito-kotlin's check function.

File: src/main/java/net/syscon/elite/api/resource/impl/OffenderMovementsResourceImpl.java [hunk 5]
```diff
@@ -18,6 +19,7 @@
 import java.time.LocalDate;
 import java.time.LocalDateTime;
 
+import static java.lang.String.valueOf;
```
Comment (andymarke) [inline src/main/java/net/syscon/elite/api/resource/impl/OffenderMovementsResourceImpl.java]: are these additions accidental?

File: src/main/java/net/syscon/elite/repository/jpa/model/BedAssignmentHistory.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    private LocalDateTime assignmentDateTime;
+
+    @Column(name = "ASSIGNMENT_REASON")
+    private String assignmentReason;
+
+    @Column(name = "ASSIGNMENT_END_DATE")
+    private LocalDate assignmentEndDate;
+
+    @Column(name = "ASSIGNMENT_END_TIME")
+    private LocalDateTime assignmentEndDateTime;
```
Comment (andymarke) [inline src/main/java/net/syscon/elite/repository/jpa/model/BedAssignmentHistory.java]: should this be LocalTime ?

File: src/test/java/net/syscon/elite/service/BedAssignmentHistoryServiceTest.java (hunk 4 above)
Comment (andymarke) [reply src/test/java/net/syscon/elite/service/BedAssignmentHistoryServiceTest.java]: to fair that looks pretty clear above

Comment (mikehalmamoj) [reply src/test/java/net/syscon/elite/service/BedAssignmentHistoryServiceTest.java]: I know but I can't use assertJ without writing my own matcher :(

File: src/main/java/net/syscon/elite/api/resource/impl/OffenderMovementsResourceImpl.java (hunk 5 above)
Comment (mikehalmamoj) [reply src/main/java/net/syscon/elite/api/resource/impl/OffenderMovementsResourceImpl.java]: Ah it was in at one point but forgot to CTRL-O.  Will do so on my next PR, thanks.

File: src/main/java/net/syscon/elite/repository/jpa/model/BedAssignmentHistory.java (hunk 6 above)
Comment (mikehalmamoj) [reply src/main/java/net/syscon/elite/repository/jpa/model/BedAssignmentHistory.java]: It's a strange thing, but the table has 2 columns which are both timestamps, but one only shows the date and the other shows both the date and time.  I'm not 100% sure what the above will do on an Oracle database so thanks for reminding me to check that out before I start using it on t3.

File: src/main/java/net/syscon/elite/service/BookingService.java [hunk 7]
```diff
@@ -666,16 +670,20 @@ public MilitaryRecords getMilitaryRecords(final Long bookingId) {
...
     @Transactional
     public void updateLivingUnit(final Long bookingId, final Long livingUnitId) {
-        final var location = locationService.getLocation(livingUnitId);
         var offenderBooking = offenderBookingRepository.findById(bookingId)
-                .orElseThrow(new EntityNotFoundException(format("Offender booking for booking id %d not found", bookingId)));
-
-        if (!offenderBooking.getLocation().getId().equals(location.getAgencyId())) {
-            throw new IllegalArgumentException(format("Move to living unit in prison %s invalid for offender in prison %s", location.getAgencyId(), offenderBooking.getLocation().getId()));
-        }
+                .orElseThrow(EntityNotFoundException.withMessage(format("Offender booking for booking id %d not found", bookingId)));
```
Comment (matt-crowson-moj) [inline src/main/java/net/syscon/elite/service/BookingService.java]: Possibly need these annotations...

`@Transactional`
`@VerifyBookingAccess`
`@HasWriteScope`

File: src/main/java/net/syscon/elite/service/BookingService.java [hunk 8]
```diff
@@ -666,16 +670,20 @@ public MilitaryRecords getMilitaryRecords(final Long bookingId) {
...
+        final var location = agencyInternalLocationRepository.findById(livingUnitId)
```
Comment (matt-crowson-moj) [inline src/main/java/net/syscon/elite/service/BookingService.java]: A general question around booking ID's.  Should we checking the booking is active?

File: src/main/java/net/syscon/elite/repository/jpa/model/BedAssignmentHistory.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    private LocalDate assignmentDate;
+
+    @Column(name = "ASSIGNMENT_TIME")
+    private LocalDateTime assignmentDateTime;
+
+    @Column(name = "ASSIGNMENT_REASON")
+    private String assignmentReason;
+
+    @Column(name = "ASSIGNMENT_END_DATE")
+    private LocalDate assignmentEndDate;
```
[the remark this answers is not in the document]
Comment (mikehalmamoj) [reply src/main/java/net/syscon/elite/repository/jpa/model/BedAssignmentHistory.java]: They are actually updated by a trigger when a new record is written... but yes I think Optional fits.  I might just do that.

File: src/main/java/net/syscon/elite/repository/jpa/model/BedAssignmentHistory.java (hunk 6 above)
Comment (Mjwillis) [reply src/main/java/net/syscon/elite/repository/jpa/model/BedAssignmentHistory.java]: Its a terrible way that its implemented - something to do with the way dates use to work in Oracle 5/6

**Pull request merged**